codec-http2/src/main/java/io/netty/handler/codec/http2/PriorityStreamByteAssigner.java

```diff
@@ -0,0 +1,437 @@
+/*
+ * Copyright 2015 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License, version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.netty.handler.codec.http2;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+
+import java.util.Arrays;
+
+/**
+ * A {@link StreamByteAssigner} that implements the HTTP/2 priority tree algorithm for assigning
+ * bytes for all streams in the connection.
+ */
+public final class PriorityStreamByteAssigner implements StreamByteAssigner {
```

Comment:
Fair points. Lets step back. What is the purpose of implementations of the `StreamByteAssigner` interface? To distribute bytes to streams to for the remote flow controller?  Will `StreamByteAssigner` ever be used without the remote flow controller? If should we make it an inner type of `Http2RemoteFlowController`? I still feel that `assign*` and `allocate*` are roughly equal in terms of having overloaded meaning in the software engineering domain. Maybe `assign*` is less overloaded in this context, but it still carries its own meaning which can lead to confusion. What about something like `ByteDistributor` (inside the scope of `Http2RemoteFlowController`)?